### ISSUE TITLE
20220227.2.1 breaks NFS mount for QNAP servers (kernel 5.16.x)

### ISSUE THREAD

**Issue by perosb** [coreos/fedora-coreos-tracker#1121]
With the kernel upgrade in **20220227.2.1** NFS fails to mount any shares from QNAP based nfs servers.

This issue is already reported in bugzilla and upstream but I'd like to raise the awareness since all my coreos based k8s nodes will fail with next stable update.

References:
https://bugzilla.redhat.com/show_bug.cgi?id=2055362
https://bugs.archlinux.org/task/73860
https://www.spinics.net/lists/linux-nfs/msg89336.html

There are some workarounds in the threads and also for archlinux the offending commit is reverted.

**Comment by dustymabe** [comment]
Reading the bug and threads, IIUC there was an incompatibility introduced in kernel 5.16.10 that causes NFSv4 clients connecting to QNAP NAS products to not work and QNAP needs to release an update to fix the problem? Also some people have reported that using NFSv3 for the time is a workaround.

Is that correct?

**Comment by perosb** [comment]
Sure, might work if downgrade to vers 3.

The same change was done in the 5.15 tree which I find odd making changes breaking.
At the same time greg states:
```
The policy for the stable kernel releases is the same as for Linus's
releases, "no user visible regressions are allowed".

There is no difference here, if something changes in one of Linus's
releases that breaks a working system, then it needs to be fixed.  The
stable kernels are not unique here at all.  Any user must be able to
always upgrade to a new kernel version without having to worry about
anything breaking.

So if there is a kernel change in Linus's tree that breaks existing
systems, it needs to be reverted or fixed to not do this.
```
https://www.spinics.net/lists/linux-nfs/msg89462.html
Not much dialog on the matter since...

I'm not very confident in QNAP fixing anything anytime soon..(I might be wrong tho)

**Comment by travier** [comment]
Per the kernel policy, this should be reverted and back ported until a fix is found. This is a user visible regression.

**Comment by achilles-git** [comment]
Hey

I see kernel-panic in 5.16.7-200.fc35.x86_64

```[  111.301380] CPU: 16 PID: 9766 Comm: mount.nfs Not tainted 5.16.7-200.fc35.x86_64 #1
[  111.302781] Hardware name: Red Hat KVM/RHEL-AV, BIOS 1.13.0-2.module_el8.5.0+2608+72063365 04/01/2014
[  111.304330] RIP: 0010:decode_getfattr_attrs+0xff0/0x1100 [nfsv4]
[  111.305541] Code: 48 8b 54 24 08 e8 10 17 33 cf 44 8b 24 24 44 89 6d 04 41 bd 00 00 00 02 44 89 75 08 41 0f cc 44 89 65 00 81 63 08 ff ff fe ff <48> 8b 55 10 48 85 d2 74 0d f6 05 f8 ef ee ff 20 0f 85 e3 92 02 00
[  111.308761] RSP: 0018:ffffb9be07e07740 EFLAGS: 00010246
[  111.309855] RAX: 0000000000000000 RBX: ffffb9be07e077bc RCX: 0000000000000000
[  111.311168] RDX: 0000000000000002 RSI: ffff9a281c45b160 RDI: ffff9a281c45b160
[  111.312481] RBP: 34f8b7efb9d7cf9d R08: 0000000000000000 R09: ffffffff90e56728
[  111.313788] R10: 000000000000000a R11: f000000000000000 R12: ffffb9be07e07800
[  111.315076] R13: 0000000000000000 R14: 000000000000001a R15: ffff9a2803d60000
[  111.316412] FS:  00007f938ad28840(0000) GS:ffff9a7d21400000(0000) knlGS:0000000000000000
[  111.317958] CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
[  111.319137] CR2: 00007f938b3e4c70 CR3: 00000002eaf16006 CR4: 0000000000770ee0
[  111.320524] DR0: 0000000000000000 DR1: 0000000000000000 DR2: 0000000000000000
[  111.321811] DR3: 0000000000000000 DR6: 00000000fffe0ff0 DR7: 0000000000000400
[  111.323066] PKRU: 55555554
[  111.323813] Call Trace:
[  111.324542]  <TASK>
[  111.325204]  decode_getfattr_generic.constprop.0+0x128/0x200 [nfsv4]
[  111.326379]  call_decode+0x1c3/0x1e0 [sunrpc]
[  111.327406]  ? rpc_decode_header+0x5c0/0x5c0 [sunrpc]
[  111.328509]  __rpc_execute+0x5f/0x440 [sunrpc]
[  111.329478]  ? __wake_up_common_lock+0x7a/0x90
[  111.330401]  rpc_execute+0xc1/0xf0 [sunrpc]
[  111.331325]  rpc_run_task+0x173/0x1c0 [sunrpc]
[  111.332299]  nfs4_call_sync_sequence+0x70/0xa0 [nfsv4]
[  111.333387]  _nfs4_proc_fs_locations+0x116/0x150 [nfsv4]
[  111.334396]  nfs4_proc_fs_locations+0x67/0xe0 [nfsv4]
[  111.335367]  nfs4_proc_lookup_common+0x1ca/0x370 [nfsv4]
[  111.336380]  nfs4_proc_lookup+0x3a/0x80 [nfsv4]
[  111.337292]  nfs_lookup+0xec/0x2a0 [nfs]
[  111.338130]  __lookup_slow+0x71/0x130
[  111.338943]  walk_component+0x11b/0x190
[  111.339782]  ? path_init+0x57/0x3f0
[  111.340528]  path_lookupat+0x6e/0x1b0
[  111.341282]  filename_lookup+0xbc/0x1a0
[  111.342053]  ? kmem_cache_alloc+0x16c/0x2d0
[  111.342850]  vfs_path_lookup+0x41/0x60
[  111.343596]  mount_subtree+0x91/0x130
[  111.344333]  ? kmem_cache_alloc_trace+0x163/0x2c0
[  111.345181]  do_nfs4_mount+0x235/0x360 [nfsv4]
[  111.345999]  nfs4_try_get_tree+0x47/0xb0 [nfsv4]
[  111.346834]  vfs_get_tree+0x22/0xb0
[  111.347515]  path_mount+0x441/0xa90
[  111.348196]  __x64_sys_mount+0xe3/0x120
[  111.348916]  do_syscall_64+0x38/0x90
[  111.349609]  entry_SYSCALL_64_after_hwframe+0x44/0xae
[  111.350574] RIP: 0033:0x7f938b3c5a4e
[  111.351263] Code: 48 8b 0d dd 73 0e 00 f7 d8 64 89 01 48 83 c8 ff c3 66 2e 0f 1f 84 00 00 00 00 00 90 f3 0f 1e fa 49 89 ca b8 a5 00 00 00 0f 05 <48> 3d 01 f0 ff ff 73 01 c3 48 8b 0d aa 73 0e 00 f7 d8 64 89 01 48
[  111.353925] RSP: 002b:00007ffe4bab99a8 EFLAGS: 00000246 ORIG_RAX: 00000000000000a5
[  111.355079] RAX: ffffffffffffffda RBX: 00007ffe4bab9b30 RCX: 00007f938b3c5a4e
[  111.356211] RDX: 0000564f0a923550 RSI: 0000564f0a923890 RDI: 0000564f0a923840
[  111.357373] RBP: 00007ffe4bab9b30 R08: 0000564f0a923b60 R09: 0033342e33332e35
[  111.358460] R10: 0000000000000000 R11: 0000000000000246 R12: 0000564f0a923b60
[  111.359548] R13: 0000000000000003 R14: 0000564f0a9239b0 R15: 0000564f09fca925
[  111.360628]  </TASK>
[  111.361150] Modules linked in: nfnetlink_queue nfnetlink_log bluetooth ecdh_generic veth nf_conntrack_netlink xt_statistic xt_recent xt_nat xt_addrtype ipt_REJECT nf_reject_ipv4 xt_CT iptable_raw rpcsec_gss_krb5 auth_rpcgss nfsv4 dns_resolver nfs lockd grace sunrpc fscache netfs iptable_mangle ip6table_mangle ip6table_filter xt_MASQUERADE xt_conntrack xt_comment iptable_filter xt_mark ip6table_nat ip6_tables iptable_nat rfkill vxlan ip6_udp_tunnel udp_tunnel nfnetlink openvswitch nsh nf_conncount nf_nat nf_conntrack nf_defrag_ipv6 nf_defrag_ipv4 overlay intel_rapl_msr intel_rapl_common isst_if_common nfit libnvdimm kvm_intel kvm irqbypass rapl iTCO_wdt intel_pmc_bxt iTCO_vendor_support cirrus drm_kms_helper joydev cec i2c_i801 i2c_smbus lpc_ich virtio_balloon drm ip_tables xfs crct10dif_pclmul crc32_pclmul crc32c_intel virtio_net ghash_clmulni_intel serio_raw net_failover virtio_blk virtio_console failover qemu_fw_cfg dm_multipath scsi_dh_rdac scsi_dh_emc scsi_dh_alua ipmi_devintf
[  111.361207]  ipmi_msghandler fuse
[  111.374210] ---[ end trace d9631d8c0966eb11 ]---
[  111.375235] RIP: 0010:decode_getfattr_attrs+0xff0/0x1100 [nfsv4]
[  111.376413] Code: 48 8b 54 24 08 e8 10 17 33 cf 44 8b 24 24 44 89 6d 04 41 bd 00 00 00 02 44 89 75 08 41 0f cc 44 89 65 00 81 63 08 ff ff fe ff <48> 8b 55 10 48 85 d2 74 0d f6 05 f8 ef ee ff 20 0f 85 e3 92 02 00
[  111.379391] RSP: 0018:ffffb9be07e07740 EFLAGS: 00010246
[  111.380381] RAX: 0000000000000000 RBX: ffffb9be07e077bc RCX: 0000000000000000
[  111.381638] RDX: 0000000000000002 RSI: ffff9a281c45b160 RDI: ffff9a281c45b160
[  111.382839] RBP: 34f8b7efb9d7cf9d R08: 0000000000000000 R09: ffffffff90e56728
[  111.384039] R10: 000000000000000a R11: f000000000000000 R12: ffffb9be07e07800
[  111.385245] R13: 0000000000000000 R14: 000000000000001a R15: ffff9a2803d60000
[  111.386442] FS:  00007f938ad28840(0000) GS:ffff9a7d21400000(0000) knlGS:0000000000000000
[  111.387777] CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
[  111.388910] CR2: 00007f938b3e4c70 CR3: 00000002eaf16006 CR4: 0000000000770ee0
[  111.390121] DR0: 0000000000000000 DR1: 0000000000000000 DR2: 0000000000000000
[  111.391439] DR3: 0000000000000000 DR6: 00000000fffe0ff0 DR7: 0000000000000400
[  111.392663] PKRU: 55555554
[  111.393433] Kernel panic - not syncing: Fatal exception
[  111.396597] Kernel Offset: 0xe000000 from 0xffffffff81000000 (relocation range: 0xffffffff80000000-0xffffffffbfffffff)
[  111.398295] Rebooting in 10 seconds..```


is there any solution for this ???

**Comment by dustymabe** [comment]
@achilles-git that's also not a kernel that's shipped in any FCOS stream right now. If you're not using FCOS this isn't the place to report issues. If you are using FCOS please open a new issue with all the details.

**Comment by jlebon** [comment]
https://forum.qnap.com/viewtopic.php?f=35&t=165385#p813893 mentions that forcing the version to 4.0 via `vers=4.0` works. Can someone hitting this issue confirm this?

**Comment by perosb** [comment]
yes, forcing version < 4.1 works.

**Comment by dustymabe** [comment]
Thank you @perosb fpr the quick feedback. I'd like to thank you once again for running the `testing` stream and helping us find issues before they hit `stable.

Unfortunately in this case we're not in a good position. The new kernel brings in an important [CVE fix](https://github.com/coreos/fedora-coreos-tracker/issues/1118), which we need to get out to users. I think the best thing we can do right now is document this problem and communicate out to the community how to workaround the issue either by showing them how freeze on an older kernel with an override or by giving them instructions to use NFS `vers=4.0`. This does require manual intervention, which is unfortunate.

**Comment by dustymabe** [comment]
Sent out an announcement for this:

- [mailing list](https://lists.fedoraproject.org/archives/list/coreos-status@lists.fedoraproject.org/thread/EEUK5UU4EFSMYJ6LW2GBLXYSFUBY5WAU/)
- [discussion forum](https://discussion.fedoraproject.org/t/latest-kernel-update-can-cause-nfs-mount-failures/37555)

**Comment by jlebon** [comment]
If using `vers=4.0` is not an option, it's also possible to pin the kernel and otherwise update everything else using e.g.:

```
[root@cosa-devsh ~]# rpm-ostree override replace https://fedoraproject-updates-archive.fedoraproject.org/fedora/35/x86_64/kernel{,-core,-modules}-5.15.18-200.fc35.x86_64.rpm
Downloading https://fedoraproject-updates-archive.fedoraproject.org/fedora/35/x86_64/kernel-5.15.18-200.fc35.x86_64.rpm...done
Downloading https://fedoraproject-updates-archive.fedoraproject.org/fedora/35/x86_64/kernel-core-5.15.18-200.fc35.x86_64.rpm...done
Downloading https://fedoraproject-updates-archive.fedoraproject.org/fedora/35/x86_64/kernel-modules-5.15.18-200.fc35.x86_64.rpm...done
Inactive base replacements:
  kernel-5.15.18-200.fc35.x86_64
  kernel-core-5.15.18-200.fc35.x86_64
  kernel-modules-5.15.18-200.fc35.x86_64
Staging deployment... done
Use "rpm-ostree override reset" to undo overrides
Run "systemctl reboot" to start a reboot
```

The "inactive" bit there is expected; there's nothing to replace currently, since the override matches the kernel version. But on upgrade, it'll kick in.

Of course, this does mean that you would not be receiving the kernel update which fixes the latest CVE or potentially future ones. So you'll want to make sure to unpin the kernel as soon as possible once it's no longer needed.

**Comment by kai-uwe-rommel** [comment]
For k8s nodes, the workaround (mount option vers=4.0) may not be practical if you are using the NFS auto provisioner.

**Comment by dustymabe** [comment]
> For k8s nodes, the workaround (mount option vers=4.0) may not be practical if you are using the NFS auto provisioner.

Hey @kai-uwe-rommel - AFAIU this problem is limited to specific products. I guess it's possible that people are using k8s nodes to talk to QNAP NAS devices via NFS. If they are not able to easily change the mount options then the solution from @jlebon in https://github.com/coreos/fedora-coreos-tracker/issues/1121#issuecomment-1067274067 will work for now.

**Comment by jmflinuxtx** [comment]
This issue seems to be limited to products no longer supported by QNAP, and the bug is on their end. Of course QNAP will not release updates for their EOL products, so it isn't going away. I have posted an update in https://bugzilla.redhat.com/show_bug.cgi?id=2055362#c33 which might be important for FCOS to pay attention to.  Specifically there is no guarantee that 5.17 kernels will have a workaround for this that doesn't require passing mount options. 

This is not a kernel bug with Fedora, or with upstream, this is a kernel bug with older kernels that QNAP is no longer updating on their end and those systems are likely exposed to a number of CVEs which have been fixed since they were EOL as well.   It is very possible that upstream just doesn't do anything here because the code is correct, and there is already a workaround to force the NFS version on the client.

**Comment by jmflinuxtx** [comment]
> https://www.spinics.net/lists/linux-nfs/msg89462.html Not much dialog on the matter since...
> 
> I'm not very confident in QNAP fixing anything anytime soon..(I might be wrong tho)

There has been considerable dialog on the matter since, but no resolution.  It just isn't a part of that thread.
https://lore.kernel.org/linux-nfs/8A1E6FE8-AD07-40B4-8765-3FFF75B465DC@oracle.com/T/#m74fd0b907a85612a414e92685492c2fd0625e69a

**Comment by perosb** [comment]
> This issue seems to be limited to products no longer supported by QNAP, and the bug is on their end. Of course QNAP will not release updates for their EOL products, so it isn't going away.

AFAIK my QNAP is supported and firmwares are being pushed, but sure, kernel version is a "bit" behind `Linux 4.2.8 #1 SMP Tue Dec 21 00:57:26 CST 2021 aarch64 GNU/Linux`. 

Thanks for the mail thread  :+1:

**Comment by jmflinuxtx** [comment]
Ahh, right. They should release an update for that one at some point then, hopefully with a fix. They aren't exactly fast, the majority of their products are on 5.10 kernels, and have not released an update for dirty pipe there either yet (your 4.2 based isn't vulnerable to that).  Several of the posts I have seen with issues are on 3.x kernels, which are not getting updates anymore.

**Comment by dustymabe** [comment]
There is a new 5.16 kernel (`kernel-5.16.15-201.fc35`) that has a fix for this. I've got a PR open to add it to our `testing-devel` branch: https://github.com/coreos/fedora-coreos-config/pull/1608

**Comment by webenhanced** [comment]
Apparently this is also happening on some Synology NASs as well.

My QNAP running QTS 5.0.0.1932 has kernel-5.10.60 also has the problem with F35 (5.16.10+), so this is not just for EOL machines.

**Comment by dustymabe** [comment]
There was a fix upstream in 5.18 that has been backported to Fedora 5.17 kernels. This means that 5.16 and 5.17+ Fedora kernels should be good. More information in [the BZ](https://bugzilla.redhat.com/show_bug.cgi?id=2055362#c58).

**Comment by dustymabe** [comment]
The fix for this went into `testing` stream release `35.20220327.2.0`. Please try out the new release and report issues.

**Comment by perosb** [comment]
`35.20220327.2.0` with `5.16.16-200.fc35.aarch64` works perfectly with vers=4.1

**Comment by webenhanced** [comment]
I can confirm that after removing the vers=4.0 from fstab and running kernel 5.16.17-200.fc35, nfs mounts perfectly on my systems.

Just adding this comment because I had reported my system having this problem above. Thanks folks.

**Comment by dustymabe** [comment]
It's my understanding that a different fix went into the `5.17` kernel for this. It would be good to get some feedback from people here for success or failure for nodes running the `next` stream.

**Comment by jmflinuxtx** [comment]
So, yes, different fix, but it isn't so much of a fix as a only do trunk discovery when a mount option is set, but this is a new mount option and it defaults to off. Testers in the original bug confirmed mounts continue to work without changing mount options.

**Comment by jmscott** [comment]
we saw an issue with qnap timeing out when the client requested the "fs_locations" attribute in version 4.2 protocol.
the client ubuntu nfs mount kept requerying the "fs_locations" attribute and not aborting.

apparently qnap was advertising support for "fs_locations" incorrectly, according to another forum.  the fix was downgrading to version 4.0 (mount option "-o minorversion=0").  since qnap appears to advertis "fs_locations" i am not sure how the query on the client could be disabled.  in others, just mount as version 4.0.

hope this helps.
